text/0000-http-index.md

```diff
@@ -63,11 +63,25 @@ Dependency files compress well. Currently the largest file of `rustc-ap-rustc_da
 
 Even in the worst case of downloading the entire index file by file, it should still use significantly less bandwidth than git clone (individually compressed files currently add up to about 39MiB).
 
+An "incremental changelog" file (described in a later section) can be used to avoid many conditional requests.
+
 ## Handling deleted crates
 
 When a client checks freshness of a crate that has been deleted, it will make a request to the server and notice a 404/410/451 HTTP status. The client can then act accordingly, and clean up local data (even tarball and source checkout).
 
-If the client is not interested in deleted crate, it won't check it, but chances are it never did, and didn't download it. If ability to proactively erase caches of deleted crates is important, then the "incremental changelog" feature can be extended to notify about deletions.
+If the client is not interested in the deleted crate, it won't check it, but chances are it never did, and didn't download it. If ability to proactively erase caches of deleted crates is important, then the "incremental changelog" feature can be extended to notify about deletions.
+
+## Dealing with inconsistent HTTP caches
+
+The index does not require all files to form one cohesive snapshot. The index is updated one file at a time. Every file is updated in a separate commit, so for every file change there exists an index state that is valid with or without it. The index only needs to preserve a partial order of updates.
+
+From Cargo's perspective dependencies are always allowed to update independently. If crate's dependencies' files are refreshed before the crate itself, it won't be different than if someone had used an older version of the crate.
+
+The only case where stale caches can cause a problem is when a new version of a crate depends on the latest version of a newly-published dependency, and caches expired for the parent crate before expiring for the dependency. Cargo will prevent that from happening, at least for the datacenter it can see. Cargo requires dependencies with sufficient versions to be already visible in the index, and won't publish a "broken" crate.
+
+Ideally, the server should ensure that a previous file change is visible everywhere before making the next change, i.e. make the CDN purge the changed file, and wait for the purge to be executed before updating files that may depend on it. This may be difficult to guarantee in a global CDNs, so Cargo needs a recovery mechanism:
+
+If a crate <var>A</var> is found to depend on a crate <var>B</var> with a version that doesn't appear to exist in the index, Cargo should fetch the crate <var>B</var> again with a cache buster. The cache buster can be a query string appended to the URL with either the current timestamp, or timestamp parsed from the `last-modified` header of the crate <var>A</var>'s response: `?cachebust=12345678`.
```

Comment:
Ah, this is interesting — I hadn't though of the problem of correlated publishes. In general, I feel like "wait a minute and retry" is an acceptable resolution for these problems, but if you're trying to publish five packages from a workspace at once, having to wait a few minutes for each one would be a huge hassle.

That said, I'm really hesitant to specifically mandate that the cache support cache busting, because it opens up an avenue for abuse. This then, to me, suggest that we may actually _need_ explicit cache invalidation, and that a TTL-based scheme will lead to pain in this relatively common circumstance. I think @Mark-Simulacrum, @pietroalbini, and @alexcrichton may also have some thoughts around this.

Reply:
This section is taken from my previous response: https://github.com/rust-lang/rfcs/pull/2789#issuecomment-637153114

Reply:
When it comes to protecting the origin server from being overwhelmed with requests from cache misses, there are two distinct scenarios:

1. Traffic from well-meaning clients that end up requesting too much
2. Intentional malicious traffic that tries to cause a Denial Of Service 

For the case 1, use of timestamps rounded to a second limits the traffic to one cache bust per second per affected crate. The need for cache bust can arise only in rare circumstances and only for a limited time after the crate has been published, so it's basically just 1 request per second on average. It won't take down even most feeble servers. And even that can be further reduced to one extra upstream request in total by using parent crate's last-modified time, so that all clients will request the same URL deterministically, regardless of their clocks.

As for the case 2, I haven't seen any CDN that would try to fight it by not contacting the origin on cache misses. CDNs deal with attacks at earlier stages by identifying abnormal traffic patterns. Layer 7 attacks are relatively expensive for attackers, and easy to block by CDNs. Even if someone did try this, there are way better URLs to attack: search (requires a DB lookup, and you can't give a strict allowlist to the CDN) or `POST` requests to signup, etc. that a CDN always has to forward.

Reply:
Do CDNs not obey cache-control headers in requests? That's what browsers normally use when the user Ctrl-F5s.

Reply:
I've tried `curl -XGET -I -Hcache-control:no-cache -Hpragma:no-cache https://static.crates.io/readmes/rand/rand-0.7.3.html` and the response still had `x-cache: Hit from cloudfront`.

Reply:
Oh, yeah, doing this with a standard HTTP header designed for the purpose would be even nicer if we could. Though that would lose the second-granularity benefit of the previous scheme. Although of course the cache could always choose to not respect the header if sent too frequently.

I do also wonder how to implement this mechanism reliably in cargo itself. Actually finding out when it's appropriate to bust is going to be pretty awkward, but I suppose that's a question for the implementation and not for the RFC.

Reply:
I feel like there is a simple and useful core to this RFC. This core pulls its weight for many uses but not all. It is great for small alternative registries. It is grate for CI. It is not grate for some publish cases with workspaces. So if we merge without the cache busting, alternative registries can use it, CI can use it, but we will need to add cache busting or content aware addressing before we default it to be on for crates.io. I am ok with that. Let us get the simple and helpful part of this idea merged in this RFC, and do a follow up RFC when we have some experience about what complexity to add.

Reply:
True, one option is to leave cache busting out of this RFC and amend it later should it be necessary. I think I'm in favor of doing it that way. Maybe mention cache busting as a "future improvement"?